File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#   def __init__(self, val=0, left=None, right=None):
#     self.val = val
#     self.left = left
#     self.right = right
class Solution:
  def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    inIdx = {}
    for i, e in enumerate(inorder):
      inIdx[e] = i
    
    def helper(postLb, postUb):
      if postLb == postUb:
        return None
      rootInIdx = inIdx[postorder[postUb-1]]
      if postUb-postLb == 1:
        return TreeNode(inorder[rootInIdx])
      l, r = postLb, postUb-2
      if inIdx[postorder[r]] < rootInIdx:
        return TreeNode(inorder[rootInIdx], helper(postLb, postUb-1), None)
      elif inIdx[postorder[l]] > rootInIdx:
        return TreeNode(inorder[rootInIdx], None, helper(postLb, postUb-1))
      mid = None
      while l <= r:
        mid = (l+r) // 2
        if inIdx[postorder[mid]] < rootInIdx and inIdx[postorder[mid+1]] > rootInIdx:
          break
        elif inIdx[postorder[mid]] < rootInIdx:
          l = mid+1
        else:
          r = mid-1
      return TreeNode(inorder[rootInIdx], helper(postLb, mid+1), helper(mid+1, postUb-1))
      
    return helper(0, len(postorder))

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *   int val;
 *   TreeNode left;
 *   TreeNode right;
 *   TreeNode() {}
 *   TreeNode(int val) { this.val = val; }
 *   TreeNode(int val, TreeNode left, TreeNode right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 *   }
 * }
 */
class Solution {
  public TreeNode helper(int postLb, int postUb, int[] inorder, int[] postorder, Map<Integer, Integer> inIdx) {
    if (postLb == postUb) return null;
    int rootInIdx = inIdx.get(postorder[postUb - 1]);
    if (postUb - postLb == 1) return new TreeNode(inorder[rootInIdx]);
    
    int l = postLb;
    int r = postUb - 2;
    if (inIdx.get(postorder[r]) < rootInIdx) {
      return new TreeNode(
        inorder[rootInIdx],
        helper(postLb, postUb - 1, inorder, postorder, inIdx),
        null
      );
    } else if (inIdx.get(postorder[l]) > rootInIdx) {
      return new TreeNode(
        inorder[rootInIdx],
        null,
        helper(postLb, postUb - 1, inorder, postorder, inIdx)
      );
    }
    int mid = (l + r) / 2;
    while (l <= r) {
      mid = (l + r) / 2;
      if (inIdx.get(postorder[mid]) < rootInIdx && inIdx.get(postorder[mid + 1]) > rootInIdx) {
        break;
      } else if (inIdx.get(postorder[mid]) < rootInIdx) {
        l = mid + 1;
      } else {
        r = mid - 1;
      }
    }
    return new TreeNode(
      inorder[rootInIdx],
      helper(postLb, mid + 1, inorder, postorder, inIdx),
      helper(mid + 1, postUb - 1, inorder, postorder, inIdx)
    );
  }
  
  public TreeNode buildTree(int[] inorder, int[] postorder) {
    Map<Integer, Integer> inIdx = new HashMap<Integer, Integer>();
    for (int i = 0; i < inorder.length; i++) {
      inIdx.put(inorder[i], i);
    }
    
    return helper(0, postorder.length, inorder, postorder, inIdx);
  }
}